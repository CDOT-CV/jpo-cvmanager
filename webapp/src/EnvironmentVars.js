class EnvironmentVars {
  static getBaseApiUrl() {
    switch (process.env.REACT_APP_ENV) {
      case 'dev':
        return process.env.REACT_APP_GATEWAY_BASE_URL_DEV
      case 'test':
        return process.env.REACT_APP_GATEWAY_BASE_URL_TEST
      case 'prod':
        return process.env.REACT_APP_GATEWAY_BASE_URL_PROD
      default:
        return process.env.REACT_APP_GATEWAY_BASE_URL
    }
  }

  static MAPBOX_TOKEN = process.env.REACT_APP_MAPBOX_TOKEN
  static KEYCLOAK_HOST_URL = process.env.REACT_APP_KEYCLOAK_URL

  static rsuInfoEndpoint = `${this.getBaseApiUrl()}/rsuinfo`
  static rsuOnlineEndpoint = `${this.getBaseApiUrl()}/rsu-online-status`
  static rsuCountsEndpoint = `${this.getBaseApiUrl()}/rsucounts`
  static rsuCommandEndpoint = `${this.getBaseApiUrl()}/rsu-command`
  static wzdxEndpoint = `${this.getBaseApiUrl()}/wzdx-feed`
  static rsuMapInfoEndpoint = `${this.getBaseApiUrl()}/rsu-map-info`
  static rsuGeoQueryEndpoint = `${this.getBaseApiUrl()}/rsu-geo-query`
  static bsmDataEndpoint = `${this.getBaseApiUrl()}/rsu-bsm-data`
  static issScmsStatusEndpoint = `${this.getBaseApiUrl()}/iss-scms-status`
  static ssmSrmEndpoint = `${this.getBaseApiUrl()}/rsu-ssm-srm-data`
  static authEndpoint = `${this.getBaseApiUrl()}/user-auth`
  static adminAddRsu = `${this.getBaseApiUrl()}/admin-new-rsu`
  static adminRsu = `${this.getBaseApiUrl()}/admin-rsu`
  static adminAddUser = `${this.getBaseApiUrl()}/admin-new-user`
  static adminUser = `${this.getBaseApiUrl()}/admin-user`
  static adminAddOrg = `${this.getBaseApiUrl()}/admin-new-org`
  static adminOrg = `${this.getBaseApiUrl()}/admin-org`
}

export default EnvironmentVars
