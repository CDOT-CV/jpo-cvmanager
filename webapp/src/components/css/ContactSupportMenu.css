#ContactSupportMenu {
    position: absolute;
    bottom: 5%;
    right: 5%;

    @media only screen and (max-height: 800px) {
        transform: translate(150%, 150%);
    }
}

h5 {
    font-size: 1.5rem;
    font-weight: 400;
    margin-bottom: 0.5rem;
    text-align: center;
}

.label {
    font-size: 1rem;
    font-weight: 400;
    margin-bottom: 0.5rem;
}

.btn {
    font-size: 1rem;
    font-weight: 400;
    margin-bottom: 0.5rem;
}

.mb-3 {
    margin-bottom: 1rem !important;
}