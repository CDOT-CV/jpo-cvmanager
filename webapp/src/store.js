import { configureStore } from "@reduxjs/toolkit";
import rsuReducer from "./slices/rsuSlice";
import userReducer from "./slices/userSlice";
import wzdxReducer from "./slices/wzdxSlice";
import configReducer from "./slices/configSlice";

const store = configureStore({
  reducer: {
    rsu: rsuReducer,
    user: userReducer,
    wzdx: wzdxReducer,
    config: configReducer,
  },
});

export default store;
