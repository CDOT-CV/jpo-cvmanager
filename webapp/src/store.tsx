import { Action, ThunkAction, configureStore } from '@reduxjs/toolkit'
import rsuReducer from './generalSlices/rsuSlice'
import userReducer from './generalSlices/userSlice'
import wzdxReducer from './generalSlices/wzdxSlice'
import configReducer from './generalSlices/configSlice'
import adminAddOrganizationReducer from './features/adminAddOrganization/adminAddOrganizationSlice'
import adminAddRsuReducer from './features/adminAddRsu/adminAddRsuSlice'
import adminAddUserReducer from './features/adminAdduser/adminAddUserSlice'
import adminEditOrganizationReducer from './features/adminEditOrganization/adminEditOrganizationSlice'
import adminEditRsuReducer from './features/adminEditRsu/adminEditRsuSlice'
import adminEditUserReducer from './features/adminEditUser/adminEditUserSlice'
import adminOrganizationTabReducer from './features/adminOrganizationTab/adminOrganizationTabSlice'
import adminOrganizationTabUserReducer from './features/adminOrganizationTabUser/adminOrganizationTabUserSlice'
import adminOrganizationTabRsuReducer from './features/adminOrganizationTabRsu/adminOrganizationTabRsuSlice'
import adminRsuTabReducer from './features/adminRsuTab/adminRsuTabSlice'
import adminUserTabReducer from './features/adminUserTab/adminUserTabSlice'
import menuReducer from './features/menu/menuSlice'
import intersectionMapReducer from './components/map/map-slice'
import intersectionMapLayerStyleReducer from './components/map/map-layer-style-slice'

export const setupStore = (preloadedState: any) => {
  return configureStore({
    reducer: {
      rsu: rsuReducer,
      user: userReducer,
      wzdx: wzdxReducer,
      config: configReducer,
      adminAddOrganization: adminAddOrganizationReducer,
      adminAddRsu: adminAddRsuReducer,
      adminAddUser: adminAddUserReducer,
      adminEditOrganization: adminEditOrganizationReducer,
      adminEditRsu: adminEditRsuReducer,
      adminEditUser: adminEditUserReducer,
      adminOrganizationTab: adminOrganizationTabReducer,
      adminOrganizationTabUser: adminOrganizationTabUserReducer,
      adminOrganizationTabRsu: adminOrganizationTabRsuReducer,
      adminRsuTab: adminRsuTabReducer,
      adminUserTab: adminUserTabReducer,
      menu: menuReducer,
      intersectionMap: intersectionMapReducer,
      intersectionMapLayerStyle: intersectionMapLayerStyleReducer,
    },
    preloadedState,
  })
}

type AppStore = ReturnType<typeof setupStore>
export type AppState = ReturnType<AppStore['getState']>
type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>

export type RootState = ReturnType<ReturnType<typeof setupStore>['getState']>
