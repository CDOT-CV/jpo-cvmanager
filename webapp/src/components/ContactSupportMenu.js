import React, { useState } from "react";
import { Form } from "react-bootstrap";
import { useForm } from "react-hook-form";

import "../components/css/Admin.css";
import "react-widgets/styles.css";
import EnvironmentVars from "../EnvironmentVars";

const ContactSupportMenu = () => {
    const [successMsg, setSuccessMsg] = useState("");
    const [errorState, setErrorState] = useState(false);
    const [errorMessage, setErrorMessage] = useState("");
    const {
        register,
        handleSubmit,
        reset,
        formState: { errors },
    } = useForm();

    const sendPostData = async (json) => {
        
        try {
            const res = await fetch(EnvironmentVars.sendEmail, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    "Authorization": "Bearer " + localStorage.getItem("token"),
                },
                body: JSON.stringify(json),
            });

            const status = res.status;
            const data = await res.json();
            if (status === 200) {
                console.debug("Successfully sent email");
                setSuccessMsg("Successfully sent email");
                setErrorState(false);
                reset();
            }
            else if (status === 500) {
                console.error(data);
                setSuccessMsg("");
                setErrorState(true);
                setErrorMessage(data);
            }
        } catch (exception_var) {
            console.error(exception_var);
            setSuccessMsg("");
            setErrorState(true);
            setErrorMessage(exception_var);
        }      
    };

    const onSubmit = (data) => {
        sendPostData(data);
    };

    return (
        <div>
            <Form onSubmit={handleSubmit(onSubmit)}>
                <h5>Contact Support</h5>
                <Form.Group className="mb-3" controlId="email">
                    <Form.Label>Your Email</Form.Label>
                    <Form.Control
                        type="email"
                        placeholder="Enter your email"
                        {...register("email", {
                            required: "Email is required",
                        })}
                    />
                    {errors.email && (
                        <Form.Text className="text-danger">
                            {errors.email.message}
                        </Form.Text>
                    )}
                </Form.Group>
                <Form.Group className="mb-3" controlId="subject">
                    <Form.Label>Subject</Form.Label>
                    <Form.Control
                        type="text"
                        placeholder="Enter your subject"
                        {...register("subject", {
                            required: "Subject is required",
                        })}
                    />
                    {errors.subject && (
                        <Form.Text className="text-danger">
                            {errors.subject.message}
                        </Form.Text>
                    )}
                </Form.Group>
                <Form.Group className="mb-3" controlId="message">
                    <Form.Label>Message</Form.Label>
                    <Form.Control
                        as="textarea"
                        rows={5}
                        placeholder="Enter your message"
                        {...register("message", {
                            required: "Message is required",
                        })}
                    />
                    {errors.message && (
                        <Form.Text className="text-danger">
                            {errors.message.message}
                        </Form.Text>
                    )}
                </Form.Group>

                {successMsg && <p className="success-msg">{successMsg}</p>}
                {errorState && (
                    <p className="error-msg">
                        Failed to send email due to error: {errorMessage}
                    </p>
                )}
                <div className="form-control">
                    <label></label>
                    <button type="submit" className="btn btn-primary">
                        Send Email
                    </button>
                </div>
            </Form>
        </div>
    );
}

export default ContactSupportMenu;