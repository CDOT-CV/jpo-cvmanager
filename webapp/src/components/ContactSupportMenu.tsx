import React, { useState } from 'react'
import { Form } from 'react-bootstrap'
import { useForm } from 'react-hook-form'

import 'react-widgets/styles.css'
import RsuApi from '../apis/rsu-api'

import './css/ContactSupportMenu.css'
import toast from 'react-hot-toast'

const ContactSupportMenu = () => {
  const [hidden, setHidden] = useState(true) // hidden by default
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm()

  const onSubmit = async (data: Object) => {
    try {
      const res = await RsuApi.postContactSupport(data)
      const status = res.status
      if (status === 200) {
        toast.success('Successfully sent email')
        reset()
      } else {
        toast.error('Something went wrong: ' + status)
      }
    } catch (exception_var) {
      toast.error('An exception occurred, please try again later')
    }
  }

  if (hidden) {
    return (
      <div>
        <button
          type="button"
          className="showbutton"
          onClick={() => {
            setHidden(!hidden)
          }}
        >
          Contact Support
        </button>
      </div>
    )
  }

  return (
    <div id="ContactSupportMenu">
      <Form onSubmit={handleSubmit(onSubmit)}>
        <h5>Contact Support</h5>
        <Form.Group className="mb-3" controlId="email">
          <Form.Label className="label">Your Email</Form.Label>
          <Form.Control
            type="email"
            placeholder="Enter your email (Required)"
            {...register('email', {
              required: 'Email is required',
            })}
          />
          {errors.email && <Form.Text className="text-danger">{errors.email.message}</Form.Text>}
        </Form.Group>
        <Form.Group className="mb-3" controlId="subject">
          <Form.Label className="label">Subject</Form.Label>
          <Form.Control
            type="text"
            placeholder="Enter your subject (Required)"
            {...register('subject', {
              required: 'Subject is required',
            })}
          />
          {errors.subject && <Form.Text className="text-danger">{errors.subject.message}</Form.Text>}
        </Form.Group>
        <Form.Group className="mb-3" controlId="message">
          <Form.Label className="label">Message</Form.Label>
          <Form.Control
            as="textarea"
            rows={5}
            placeholder="Enter your message (Required)"
            {...register('message', {
              required: 'Message is required',
            })}
          />
          {errors.message && <Form.Text className="text-danger">{errors.message.message}</Form.Text>}
        </Form.Group>
        <div className="form-control">
          <label></label>
          <button type="submit" className="btn btn-primary">
            Send Email
          </button>
        </div>
        <div>
          <button
            type="button"
            className="hidebutton"
            onClick={() => {
              setHidden(!hidden)
            }}
          >
            x
          </button>
        </div>
      </Form>
    </div>
  )
}

export default ContactSupportMenu
