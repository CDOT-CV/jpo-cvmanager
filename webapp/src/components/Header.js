import React, { useState, useEffect } from 'react'
import Grid from '@material-ui/core/Grid'
import { GoogleLogin } from '@react-oauth/google'
import '../features/menu/Menu.js'
import logo from '../images/cdot_logo.png'

import { useSelector, useDispatch } from 'react-redux'
import {
  selectOrganizationName,
  selectName,
  selectEmail,
  selectAuthLoginData,
  selectTokenExpiration,
  selectLoginFailure,

  // actions
  login,
  logout,
  changeOrganization,
  setLoginFailure,
  selectLoginMessage,
  setLoginMessage
} from '../generalSlices/userSlice'

import './css/Header.css'

import ContactSupportMenu from "./ContactSupportMenu";

const Header = (props) => {
  const dispatch = useDispatch()

  const authLoginData = useSelector(selectAuthLoginData)
  const organizationName = useSelector(selectOrganizationName)
  const userName = useSelector(selectName)
  const userEmail = useSelector(selectEmail)
  const tokenExpiration = useSelector(selectTokenExpiration)
  const loginFailure = useSelector(selectLoginFailure)
  const loginMessage = useSelector(selectLoginMessage)


  const [tokenExpired, setTokenExpired] = useState(false)

  useEffect(() => {
    setLoginFailure(!authLoginData)
  }, [authLoginData])

  useEffect(() => {
    setTokenExpired(Date.now() < tokenExpiration)
  }, [tokenExpiration])

  return (
    <div>
      {authLoginData ? (
        <header id="header">
          <Grid container alignItems="center">
            <img id="logo" src={logo} alt="Logo" />
            <h1 id="header-text">CDOT CV Manager</h1>
            <div id="login">
              <Grid container alignItems="center">
                <Grid id="userInfoGrid">
                  <h3 id="nameText">{userName}</h3>
                  <h3 id="emailText">{userEmail}</h3>
                  <select
                    id="organizationDropdown"
                    value={organizationName}
                    onChange={(event) => dispatch(changeOrganization(event.target.value))}
                  >
                    {(authLoginData?.data?.organizations ?? []).map((permission) => (
                      <option key={permission.name + 'Option'} value={permission.name}>
                        {permission.name} ({permission.role})
                      </option>
                    ))}
                  </select>
                </Grid>
                <button id="logout" onClick={() => dispatch(logout())}>
                  Logout
                </button>
              </Grid>
            </div>
          </Grid>
        </header>
      ) : (
        <div id="frontpage">
          <Grid container id="frontgrid" alignItems="center" direction="column">
            <Grid container justifyContent="center" alignItems="center">
              <img id="frontpagelogo" src={logo} alt="Logo" />
              <h1 id="header-text">CDOT CV Manager</h1>
            </Grid>
            <div id="googlebtn">
              <GoogleLogin
                onSuccess={(res) => dispatch(login(res))}
                onError={(err) => dispatch(setLoginMessage('Google Login Error: ' + err))}
                text="signin_with"
                size="large"
                theme="outline" />
            </div>
            {loginFailure && <h3 id="loginMessage">{ loginMessage }</h3>}
            {tokenExpired && <h3 id="loginMessage">Login Timed Out</h3>}

            <br />
            <ContactSupportMenu />
          </Grid>
        </div>
      )}
    </div>
  )
}

export default Header
