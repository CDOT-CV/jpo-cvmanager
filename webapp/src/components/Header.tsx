import React, { useEffect } from 'react'
import Grid from '@material-ui/core/Grid'
import logo from '../images/logo.png'
import { useSelector, useDispatch } from 'react-redux'
import { DotName } from '../constants'
import {
  selectOrganizationName,
  selectName,
  selectEmail,
  selectAuthLoginData,
  selectLoginFailure,
  selectKcFailure,

  // actions
  logout,
  changeOrganization,
  setLoginFailure,
  setKcFailure,
} from '../generalSlices/userSlice'
import { useKeycloak } from '@react-keycloak/web'

import './css/Header.css'

import ContactSupportMenu from './ContactSupportMenu'
import { AnyAction, ThunkDispatch } from '@reduxjs/toolkit'
import { RootState } from '../store'

const Header = () => {
  const dispatch: ThunkDispatch<RootState, void, AnyAction> = useDispatch()
  const { keycloak } = useKeycloak()

  const authLoginData = useSelector(selectAuthLoginData)
  const organizationName = useSelector(selectOrganizationName)
  const userName = useSelector(selectName)
  const userEmail = useSelector(selectEmail)
  const loginFailure = useSelector(selectLoginFailure)
  const kcFailure = useSelector(selectKcFailure)

  useEffect(() => {
    dispatch(setLoginFailure(!authLoginData))
  }, [authLoginData])

  useEffect(() => {
    if (!keycloak?.authenticated) {
      const timer = setTimeout(() => {
        console.debug('Login failure logic: User is not authenticated with keycloak')
        dispatch(setKcFailure(true))
      }, 590000)
      return () => clearTimeout(timer)
    } else {
      console.debug('Login failure logic: User is now authenticated with keycloak')
      dispatch(setKcFailure(false))
    }
  }, [keycloak, keycloak?.authenticated, dispatch])

  const handleUserLogout = () => {
    console.debug('handleUserLogout')
    dispatch(logout())
    keycloak?.logout()
  }

  return (
    <div>
      {authLoginData && keycloak?.authenticated ? (
        <header id="header">
          <Grid container alignItems="center">
            <img id="logo" src={logo} alt="Logo" />
            <h1 id="header-text">{DotName} CV Manager</h1>
            <div id="login">
              <Grid container alignItems="center">
                <Grid id="userInfoGrid">
                  <h3 id="nameText">{userName}</h3>
                  <h3 id="emailText">{userEmail}</h3>
                  <select
                    id="organizationDropdown"
                    value={organizationName}
                    onChange={(event) => dispatch(changeOrganization(event.target.value))}
                  >
                    {(authLoginData?.data?.organizations ?? []).map((permission) => (
                      <option key={permission.name + 'Option'} value={permission.name}>
                        {permission.name} ({permission.role})
                      </option>
                    ))}
                  </select>
                </Grid>
                <button id="logout" onClick={() => handleUserLogout()}>
                  Logout
                </button>
              </Grid>
            </div>
          </Grid>
        </header>
      ) : (
        <div id="frontpage">
          <Grid container id="frontgrid" alignItems="center" direction="column">
            <Grid container justifyContent="center" alignItems="center">
              <img id="frontpagelogo" src={logo} alt="Logo" />
              <h1 id="header-text">{DotName} CV Manager</h1>
            </Grid>

            <div id="keycloakbtndiv">
              {keycloak?.authenticated && (
                <button className="keycloak-button" onClick={() => handleUserLogout()}>
                  Logout User
                </button>
              )}
            </div>
            {loginFailure && <h3 id="loginMessage">User Unauthorized, Please Contact Support</h3>}
            {kcFailure && <h3 id="loginMessage">Application Authentication Error!</h3>}

            <br />
            <ContactSupportMenu />
          </Grid>
        </div>
      )}
    </div>
  )
}

export default Header
