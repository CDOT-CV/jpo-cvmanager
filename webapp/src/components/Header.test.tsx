import React from 'react'
import { render } from '@testing-library/react'
import Header from './Header'
import { Provider } from 'react-redux'
import { setupStore } from '../store'
import EnvironmentVars from '../EnvironmentVars'
import { useKeycloak } from '@react-keycloak/web'
import { replaceChaoticIds } from '../utils/test-utils'
import ContactSupportMenu from './ContactSupportMenu'

jest.mock('@react-keycloak/web')

const mockKeycloak = {
  authenticated: true,
  login: jest.fn(),
  logout: jest.fn(),
  register: jest.fn(),
  accountManagement: jest.fn(),
  loadUserProfile: jest.fn(),
}

describe('<Header />', () => {
  beforeEach(() => {
    ;(useKeycloak as any).mockReturnValue([mockKeycloak])
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should take a snapshot', () => {
    const { container } = render(
      <Provider
        store={setupStore({
          user: {
            value: {
              loginFailure: true,
              loginMessage: 'User Unauthorized',
              atuhLoginData: {
                data: {
                  organizations: [{ name: 'org1', role: 'role1' }],
                },
              },
            },
          },
        })}
      >
        <Header />
        <br />
        <ContactSupportMenu />
      </Provider>
    )

    expect(replaceChaoticIds(container)).toMatchSnapshot()
  })
})
