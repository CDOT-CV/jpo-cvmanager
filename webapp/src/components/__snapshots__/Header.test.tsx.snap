// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should take a snapshot 1`] = `
<div>
  <div>
    <div
      id="frontpage"
    >
      <div
        class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-column css-mocked-MuiGrid2-root"
        id="frontgrid"
      >
        <div
          class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row css-mocked-MuiGrid2-root"
        >
          <img
            alt="Logo"
            id="frontpagelogo"
            src="logo.png"
          />
          <h1
            id="header-text"
          >
             CV Manager
          </h1>
        </div>
        <h3
          id="loginMessage"
        >
          User Unauthorized
        </h3>
        <div
          id="keycloakbtndiv"
        >
          <button
            class="keycloak-button"
          >
            Logout User
          </button>
        </div>
        <br />
        <div>
          <button
            class="showbutton"
            type="button"
          >
            Contact Support
          </button>
        </div>
      </div>
    </div>
  </div>
  <br />
  <div>
    <button
      class="showbutton"
      type="button"
    >
      Contact Support
    </button>
  </div>
</div>
`;
